File: app/src/main/java/org/wikiedufoundation/wikiedudashboard/course_detail/provider/RetrofitCourseDetailProvider.java
package org.wikiedufoundation.wikiedudashboard.course_detail.provider;

import android.util.Log;

import org.wikiedufoundation.wikiedudashboard.course_detail.data.CourseDetailResponse;
import org.wikiedufoundation.wikiedudashboard.helper.PresenterCallback;
import org.wikiedufoundation.wikiedudashboard.helper.ProviderUtils;
import org.wikiedufoundation.wikiedudashboard.helper.WikiEduDashboardApi;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class RetrofitCourseDetailProvider implements CourseDetailProvider {

    private WikiEduDashboardApi wikiEduDashboardApi;
    public RetrofitCourseDetailProvider() {
        wikiEduDashboardApi = ProviderUtils.getApiObject();
    }

    @Override
    public void requestCourseDetail(String url, final PresenterCallback presenterCallback) {
        String sub_url = "courses/" + url + "/course.json";
        Call<CourseDetailResponse> courseDetailResponseCall = wikiEduDashboardApi.getCourseDetail(sub_url);
        courseDetailResponseCall.enqueue(new Callback<CourseDetailResponse>() {
            @Override
            public void onResponse(Call<CourseDetailResponse> call, Response<CourseDetailResponse> response) {
                Log.d("Success: ",response.body().getCourse()+"");
                presenterCallback.onSuccess(response.body());
            }

            @Override
            public void onFailure(Call<CourseDetailResponse> call, Throwable t) {
                presenterCallback.onFailure();
                t.printStackTrace();
                Log.d("Failure: ",t.getMessage()+"");
            }
        });
    }
}

File: app/src/main/java/org/wikiedufoundation/wikiedudashboard/ui/coursedetail/common/provider/RetrofitCourseDetailProvider.kt
package org.wikiedufoundation.wikiedudashboard.ui.coursedetail.common.provider

import org.wikiedufoundation.wikiedudashboard.data.network.WikiEduDashboardApi
import org.wikiedufoundation.wikiedudashboard.ui.coursedetail.common.data.CourseDetailResponse
import org.wikiedufoundation.wikiedudashboard.util.PresenterCallback
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import timber.log.Timber

/**
 * Retrofit http request to get course detail data
 * ***/
class RetrofitCourseDetailProvider(
        private val wikiEduDashboardApi: WikiEduDashboardApi
) : CourseDetailProvider {
//     = ProviderUtils.apiObject

    override fun requestCourseDetail(url: String, presenterCallback: PresenterCallback<*>) {
        val subUrl = "courses/$url/course.json"
        val courseDetailResponseCall = wikiEduDashboardApi.getCourseDetail(subUrl)
        courseDetailResponseCall.enqueue(object : Callback<CourseDetailResponse> {
            override fun onResponse(call: Call<CourseDetailResponse>, response: Response<CourseDetailResponse>) {
                Timber.d("${response.body().toString()} ")
                presenterCallback.onSuccess(response.body())
            }

            override fun onFailure(call: Call<CourseDetailResponse>, t: Throwable) {
                presenterCallback.onFailure()
                t.printStackTrace()
                Timber.d("${t.message} ")
            }
        })
    }
}
